### PR TITLE
feat: add NVIDIA Brev provider

### LINKED ISSUE #289: Feature request: add NVIDIA Brev provider for GPU-backed Crabbox leases
# Feature request: add NVIDIA Brev provider for GPU-backed Crabbox leases

## Summary

Add a first-class Crabbox provider for [NVIDIA Brev](https://docs.nvidia.com/brev) so Crabbox can provision Brev GPU instances, resolve their SSH access, run normal Crabbox sync/run/ssh workflows, and release the instances when the lease is done.

Suggested provider name: `nvidia-brev`

Suggested aliases: `brev`, `nvidia`

## Motivation

Crabbox already has strong provider abstractions for remote Linux execution, including GPU-oriented SSH lease providers such as RunPod and delegated cloud sandbox providers. NVIDIA Brev is a natural fit for the same workflow because it provides preconfigured GPU instances for AI/ML development across multiple cloud providers, with NVIDIA drivers, CUDA, Python, Docker, Jupyter, CLI management, SSH access, and port forwarding already built into the platform.

The user workflow I want:

```sh
crabbox run --provider nvidia-brev -- nvidia-smi
crabbox run --provider nvidia-brev -- python -c 'import torch; print(torch.cuda.is_available())'
crabbox ssh --provider nvidia-brev
crabbox stop --provider nvidia-brev <lease-id-or-slug>
```

This would let Crabbox users run GPU-dependent tests, model-training smoke checks, CUDA/PyTorch validation, Docker/NVIDIA Container Toolkit checks, and agent workloads without manually creating and tearing down Brev machines outside Crabbox.

## Local evidence from investigation

I inspected the local Crabbox checkout, the local `brev-cli` checkout, and the local NVIDIA Brev docs snapshot in this workspace.

Relevant Crabbox evidence:

- Crabbox provider docs define `SSHLeaseBackend` as the right shape when a provider can hand Crabbox a reachable SSH target. Core then owns sync, command execution, result collection, status rendering, and release workflows.
- `internal/providers/runpod` is the closest existing GPU provider pattern: direct-only Linux SSH lease provider, public SSH target, `doctor`, `list`, `status`, `stop`, provider-specific config/flags, and cost-discipline docs.
- Crabbox already supports SSH proxy-style targets through `SSHTarget.ProxyCommand`; `internal/providers/tenki` uses that path today, and the shared SSH/rsync code passes `-o ProxyCommand=...` into both SSH and rsync.
- New built-in providers live under `internal/providers/<name>` and are side-effect registered from `internal/providers/all/all.go`.

Relevant NVIDIA Brev docs evidence:

- Brev provides GPU instances across multiple cloud providers, preconfigured with NVIDIA drivers, CUDA, Python, Docker, Jupyter, and AI/ML tooling.
- Brev supports headless CLI auth with `brev login --token`; credentials and SSH keys live under `~/.brev/`.
- The documented lifecycle includes `brev create`, `brev list`, `brev start`, `brev stop`, `brev delete`, and `brev refresh`.
- `brev create` supports GPU-oriented selection through `--type`, `--gpu-name`, provider/search filters, fallback chains, startup scripts, launchables, dry-run, detached mode, and count/parallel creation.
- `brev list --json` is documented and the local CLI implements it.
- `brev refresh` updates `~/.brev/ssh_config` so users can connect by instance name.
- Connectivity can be direct SSH or proxy-based. The local CLI writes SSH config entries that may contain `HostName`/`Port` or `ProxyCommand` entries, including Cloudflare-backed SSH proxy commands.

Relevant Brev CLI code evidence:

- `pkg/cmd/gpucreate` implements the current `brev create` command with `--type`, `--gpu-name`, `--provider`, `--startup-script`, `--mode`, `--container-image`, `--compose-file`, `--launchable`, `--dry-run`, `--detached`, `--count`, and `--parallel`.
- `pkg/cmd/ls` implements `brev ls --json`, returning workspace names, ids, statuses, instance type/kind, GPU label, build status, shell status, and health status.
- `pkg/cmd/stop` and `pkg/cmd/delete` implement lifecycle shutdown/deletion.
- `pkg/cmd/shell`, `pkg/cmd/exec`, and `pkg/cmd/portforward` prove Brev instances are already designed for SSH command execution and port forwarding.
- `pkg/ssh/sshconfigurer.go` writes direct `HostName`/`Port` SSH entries and `ProxyCommand` SSH entries, both with `IdentityFile ~/.brev/brev.pem`-style key material.

Local host checks after installing and authenticating the Brev CLI:

- `brev --version` reports `Current Version: v0.6.326`.
- `brev --help`, `brev create --help`, and `brev ls --help` confirm the command/flag surface above on the installed binary.
- After `brev login`, `brev ls --json` succeeds in the active org and returns `{"workspaces": null}` when no instances exist.
- `brev ls orgs --json` succeeds and reports the active organization; the local draft intentionally does not include the account-specific org name or id.
- `brev search gpu --gpu-name A100 --json` succeeds and returns machine-readable GPU catalog entries.
- `brev create crabbox-dry-run --gpu-name A100 --dry-run` succeeds without creating an instance and returns the fallback instance-type list that would be attempted.
- `brev refresh` succeeds. On this machine, only credential/settings files exist under `~/.brev/` while no instances exist; `~/.brev/ssh_config` and key material should therefore be treated as post-instance/refresh artifacts, not unconditional post-login artifacts.

I did not find a stable public REST API contract in the local NVIDIA docs snapshot. Because of that, the first Crabbox implementation should probably integrate through the installed `brev` CLI plus its generated SSH config, unless a stable documented API is identified during implementation.

## Proposed MVP

Implement `provider: nvidia-brev` as a direct-only Linux SSH lease provider.

Provider spec:

- Kind: `ProviderKindSSHLease`
- Target: Linux only
- Features: `ssh`, `crabbox-sync`
- Coordinator: never/direct-only for the first version
- Auth: use existing Brev CLI auth state; do not add token/API-key command-line flags

Expected config shape:

```yaml
provider: nvidia-brev
target: linux
nvidiaBrev:
  cli: brev
  org: ""
  type: ""
  gpuName: "A100"
  provider: ""
  mode: vm
  launchable: ""
  startupScript: ""
  releaseAction: delete # delete | stop
  target: container     # container | host, if Brev SSH config exposes both
  user: ""
  workRoot: /tmp/crabbox
```

Expected environment overrides:

```text
CRABBOX_NVIDIA_BREV_CLI
CRABBOX_NVIDIA_BREV_ORG
CRABBOX_NVIDIA_BREV_TYPE
CRABBOX_NVIDIA_BREV_GPU_NAME
CRABBOX_NVIDIA_BREV_PROVIDER
CRABBOX_NVIDIA_BREV_MODE
CRABBOX_NVIDIA_BREV_LAUNCHABLE
CRABBOX_NVIDIA_BREV_STARTUP_SCRIPT
CRABBOX_NVIDIA_BREV_RELEASE_ACTION
CRABBOX_NVIDIA_BREV_TARGET
CRABBOX_NVIDIA_BREV_USER
CRABBOX_NVIDIA_BREV_WORK_ROOT
```

Expected flags:

```text
--nvidia-brev-cli
--nvidia-brev-org
--nvidia-brev-type
--nvidia-brev-gpu-name
--nvidia-brev-cloud-provider
--nvidia-brev-mode
--nvidia-brev-launchable
--nvidia-brev-startup-script
--nvidia-brev-release-action
--nvidia-brev-target
--nvidia-brev-user
--nvidia-brev-work-root
```

No Brev token, API key, refresh token, or password should be accepted as a Crabbox flag. The user should authenticate with Brev through `brev login` or `brev login --token` before using the provider.

### Doctor

`crabbox doctor --provider nvidia-brev` should be read-only.

It should verify:

- the configured Brev CLI is on `PATH`;
- `brev --version` succeeds;
- Brev auth is usable, preferably through `brev ls --json`;
- `~/.brev/` exists with credentials/settings after auth;
- `~/.brev/ssh_config` and key material exist only when Brev has generated them for SSH-capable instances; absence of those files in an empty account should not fail doctor unless a specific existing instance is being resolved;
- the selected org can be resolved if `nvidiaBrev.org` is set;
- if practical, the requested GPU/type filters can be checked with Brev search or dry-run without creating an instance.

Doctor must not create, start, stop, or delete a Brev instance.

### Acquire

`warmup` / `run` without an id should:

1. Allocate a Crabbox lease id and slug.
2. Create a Brev instance with a Crabbox-owned name, for example `crabbox-<slug>-<lease-suffix>`.
3. Pass through configured type/GPU/provider/mode/launchable/startup-script options.
4. Wait until Brev reports the instance as running/ready.
5. Run `brev refresh` so the local SSH config is current.
6. Resolve the matching SSH entry from `~/.brev/ssh_config`.
7. Populate a Crabbox `SSHTarget` from the SSH config:
   - `User`
   - `Host`
   - `Port`
   - `Key`
   - `ProxyCommand`, when present
   - `KnownHostsFile`, if a provider-specific known-hosts boundary is needed
8. Run Crabbox's normal SSH readiness check.
9. Ensure required remote tools exist (`git`, `rsync`, `tar`, and a shell); install or clearly fail with a remediation hint if Brev's image lacks them.
10. Store a local Crabbox claim that includes the Brev workspace id/name and the SSH alias used.

### Resolve, list, and status

The provider should support normal Crabbox resolution by:

- Crabbox lease id;
- Crabbox slug;
- raw Brev instance/workspace id, when available;
- Brev instance name, when it is unambiguous.

`crabbox list --provider nvidia-brev` should use Brev inventory and filter to Crabbox-owned instances by default. `--all` can show all accessible Brev instances in the configured org.

`crabbox status --provider nvidia-brev --id <id>` should normalize Brev states into Crabbox status output while preserving useful provider details such as Brev id, Brev name, GPU/type, and current state.

### Release and cleanup

Because GPU instances can be expensive, the provider needs explicit cost discipline.

Default recommendation:

- `releaseAction: delete` for ephemeral Crabbox leases, so `crabbox stop` deletes the Brev instance.
- `releaseAction: stop` as an opt-in when the user wants data persistence or faster restart.

`crabbox cleanup --provider nvidia-brev --dry-run` should list stale Crabbox-owned Brev instances before acting. Non-dry-run cleanup should only affect instances that have both:

- a Crabbox-owned name or claim; and
- a stale/expired Crabbox ownership marker that belongs to this repo/provider scope.

It must not stop or delete arbitrary Brev instances that happen to be visible in the account.

## Acceptance criteria

- `crabbox providers --json` lists `nvidia-brev` with aliases and Linux SSH/sync capabilities.
- `crabbox doctor --provider nvidia-brev` performs only read-only checks and gives actionable errors when `brev` is missing, auth is missing, or the requested org/config cannot be resolved.
- `crabbox warmup --provider nvidia-brev` can create a GPU instance, refresh Brev SSH config, resolve a direct or proxy-backed SSH target, and store a Crabbox local claim.
- `crabbox run --provider nvidia-brev -- nvidia-smi` runs on the Brev instance using Crabbox's normal sync and SSH execution path.
- `crabbox ssh --provider nvidia-brev --id <id>` resolves a kept Brev lease and opens SSH through the correct direct or `ProxyCommand` route.
- `crabbox list --provider nvidia-brev` shows Crabbox-owned Brev leases by default; `--all` can show all visible Brev instances without implying ownership.
- `crabbox status --provider nvidia-brev --id <id>` reports normalized Crabbox state plus useful Brev provider metadata.
- `crabbox stop --provider nvidia-brev <id>` applies the configured release action and removes the local claim only after the provider-side action succeeds.
- Failed acquire rolls back the Brev instance when `--keep` is not set.
- No Brev token, API key, refresh token, org id, or private SSH key is printed in normal logs, config-show output, errors, test fixtures, or public docs.

## Tests and verification expected in a PR

Unit tests should cover:

- provider registration, aliases, spec, target rejection, and feature flags;
- config file loading and environment override loading;
- provider-specific flag handling;
- missing Brev CLI/auth doctor failures;
- `brev ls --json` parsing, including `{"workspaces": null}`, empty arrays, populated arrays, and malformed output;
- Brev SSH config parsing for direct `HostName`/`Port` entries;
- Brev SSH config parsing for `ProxyCommand` entries;
- acquire command construction without putting secrets on argv;
- rollback on failed acquire;
- resolve by lease id, slug, Brev id, and Brev name;
- list filtering for Crabbox-owned instances versus `--all`;
- release actions `delete` and `stop`;
- cleanup dry-run safety, including "skip unclaimed instance" behavior.

Suggested focused verification commands:

```sh
go test ./internal/providers/nvidiabrev ./internal/providers/all ./internal/cli ./cmd/crabbox
go test -race ./internal/providers/nvidiabrev ./internal/providers/all ./internal/cli
go vet ./...
node scripts/check-docs-links.mjs
node scripts/check-command-docs.mjs
```

A live smoke script would be useful but should be opt-in and credential-gated, for example:

```sh
CRABBOX_NVIDIA_BREV_LIVE=1 scripts/live-nvidia-brev-smoke.sh
```

The live smoke should create a small/cheap configured GPU instance, run `nvidia-smi`, prove Crabbox sync/run works, then release/delete the instance according to the configured release action.

## Documentation updates

A PR should add:

- `docs/providers/nvidia-brev.md`
- provider entry in `docs/provider-backends.md`
- command/config examples wherever provider docs are indexed
- live smoke docs for any `scripts/live-nvidia-brev-smoke.sh`

The provider docs should clearly explain:

- users authenticate with `brev login` or `brev login --token`;
- Crabbox does not accept Brev secrets as flags;
- `releaseAction: delete` versus `releaseAction: stop`;
- Brev may use direct SSH or `ProxyCommand`/Cloudflare-backed SSH;
- `brev refresh` is part of the provider flow;
- which Brev target is used by default when both container and host SSH aliases exist;
- which remote tools Crabbox expects on the instance.

## Open questions

- Is there a stable, documented Brev API that Crabbox should call directly, or should the first implementation intentionally shell out to `brev` and parse `brev ls --json` plus `~/.brev/ssh_config`?
- Should the default execution target be Brev's container SSH entry or host SSH entry? Container mode likely matches `brev shell`, but host mode may be better for Docker/NVIDIA Container Toolkit workflows.
- Does `brev create` expose the new instance id in a stable machine-readable way, or should Crabbox resolve by generated name after creation?
- Should the provider support Brev launchables in the first PR or leave `nvidiaBrev.launchable` for a follow-up?
- Are there Brev instance types that lack `rsync` or `tar`, and if so should Crabbox bootstrap them or fail with a concise setup hint?

## References inspected locally

- `crabbox/docs/provider-backends.md`
- `crabbox/internal/providers/runpod`
- `crabbox/internal/providers/tenki`
- `crabbox/internal/cli/ssh.go`
- `brev-cli/pkg/cmd/gpucreate`
- `brev-cli/pkg/cmd/ls`
- `brev-cli/pkg/cmd/stop`
- `brev-cli/pkg/cmd/delete`
- `brev-cli/pkg/cmd/shell`
- `brev-cli/pkg/cmd/exec`
- `brev-cli/pkg/cmd/portforward`
- `brev-cli/pkg/ssh/sshconfigurer.go`
- `docs.nvidia.com/2026-06-13_docs_nvidia_com_combined.md`

### PR DESCRIPTION
**Pull request by coygeek**
Closes openclaw/crabbox#289

## Summary
- add the `nvidia-brev` provider as a Linux SSH-lease backend using local Brev CLI auth
- implement config/env/flag plumbing, provider registration, lifecycle operations, SSH config resolution, local claim handling, release, cleanup, and doctor/list behavior
- add provider docs plus a standalone opt-in live smoke harness gated by `CRABBOX_NVIDIA_BREV_LIVE=1`

## Verification
- `go test ./internal/providers/nvidiabrev ./internal/providers/all ./internal/cli ./cmd/crabbox`
- `go test -race ./internal/providers/nvidiabrev ./internal/providers/all ./internal/cli`
- `go vet ./...`
- `node scripts/check-docs-links.mjs`
- `node scripts/check-command-docs.mjs`
- `node --test scripts/live-nvidia-brev-smoke.test.js`
- `scripts/live-nvidia-brev-smoke.sh` -> `classification=environment_blocked reason=CRABBOX_NVIDIA_BREV_LIVE_not_enabled`
- `git diff --check`
- `rg -n "nvidia-brev-(token|api-key|password|private-key)|NVIDIA_BREV_(TOKEN|API_KEY|PASSWORD|PRIVATE_KEY)" internal docs scripts` matched only intentional negative tests / secret-surface assertions
- `$HOME/.agents/skills/autoreview/scripts/autoreview --mode branch --base origin/main --engine codex --model gpt-5.5 --thinking medium` -> clean, no accepted/actionable findings

## Notes
The live Brev smoke is intentionally not run by default because it can create billable GPU resources. It requires explicit local opt-in and uses local Brev CLI credentials rather than accepting Crabbox-managed secret flags.

### CODE REVIEW

**Comment by steipete** [comment]
Maintainer verification update:

- Full branch autoreview: clean, no actionable findings.
- Local Go gate: `go vet ./...` and `go test -race ./...` passed.
- Worker gate: format, lint, typecheck, 468 tests, and dry-run build passed.
- Docs gate: 47 command docs, 51-provider matrix, 173 Markdown files, and docs-site build passed.
- NVIDIA Brev smoke harness: 7/7 tests passed.
- GitHub CI: all 8 checks passed.

Real live E2E is still blocked before resource creation. Brev CLI v0.6.326 returns `ErrorForbidden`; the authenticated Brev web session is gated by a Terms of Service acceptance modal. No GPU workspace or billable resource was created. Leaving this open until an account owner accepts the terms and the guarded `nvidia-smi` lifecycle smoke passes.

**Comment by steipete** [comment]
Live E2E passed on June 13, 2026.

Proof:
- Authenticated Brev CLI v0.6.326 through NVIDIA KAS.
- Active organization verified before lifecycle operations.
- Guarded smoke provisioned one `hyperstack_A100_80G` workspace.
- Crabbox waited for Brev build and shell readiness, refreshed generated SSH config, connected over SSH, and ran `nvidia-smi` successfully.
- Crabbox list validation passed.
- Smoke deleted the workspace and confirmed both `brev ls --json` and `crabbox list --provider nvidia-brev --all --json` were empty afterward.
- Stable result: `classification=live_nvidia_brev_smoke_passed`.

Autoreview is clean and all 8 required checks are green. Landing.
